File: apps/web-app/src/lib/socket-client.ts
import { io, Socket } from 'socket.io-client';
import { CHAT_SOCKET_EVENTS, CHAT_SOCKET_NAMESPACE } from '@urban/shared-constants';
import { readAccessToken } from './api-client';

const SOCKET_URL = import.meta.env.VITE_SOCKET_URL || 'http://localhost:3001';

class SocketClient {
  public socket: Socket | null = null;
  private connectPromise: Promise<void> | null = null;
  private authToken: string | null = null;

  connect(): Promise<void> {
    if (this.connectPromise) return this.connectPromise;

    this.connectPromise = new Promise((resolve, reject) => {
      try {
        const token = readAccessToken();

        if (this.socket?.connected && this.authToken === token) {
          this.connectPromise = null;
          resolve();
          return;
        }

        if (this.socket && this.authToken === token) {
          const onConnect = () => {
            this.socket?.off('connect_error', onConnectError);
            this.connectPromise = null;
            resolve();
          };

          const onConnectError = (error: unknown) => {
            this.socket?.off('connect', onConnect);
            this.connectPromise = null;
            reject(error);
          };

          this.socket.once('connect', onConnect);
          this.socket.once('connect_error', onConnectError);
          this.socket.connect();
          return;
        }

        if (this.socket && this.authToken !== token) {
          this.socket.removeAllListeners();
          this.socket.disconnect();
          this.socket = null;
        }

        this.socket = io(`${SOCKET_URL}${CHAT_SOCKET_NAMESPACE}`, {
          transports: ['websocket'],
          auth: { token },
          reconnectionAttempts: 5,
          timeout: 10000,
        });
        
        this.authToken = token;

        this.socket.on('connect', () => {
          console.log('[Web Socket] Connected');
          this.connectPromise = null;
          resolve();
        });

        this.socket.on(CHAT_SOCKET_EVENTS.READY, () => {
          console.log('[Web Socket] Chat Socket Ready');
        });

        this.socket.on('disconnect', (reason) => {
          console.log('[Web Socket] Disconnected:', reason);
          this.connectPromise = null;
        });

        this.socket.on('connect_error', (error) => {
          console.error('[Web Socket] Connect error:', error);

          this.connectPromise = null;
          reject(error);
        });
      } catch (e) {
        console.error('[Web Socket] Failed to init socket', e);
        this.connectPromise = null;
        reject(e);
      }
    });

    return this.connectPromise;
  }

  disconnect() {
    if (this.socket) {
      this.socket.removeAllListeners();
      this.socket.disconnect();
      this.socket = null;
    }
    this.authToken = null;
    this.connectPromise = null;
  }

  emitWithAck<T = any>(event: string, payload: any): Promise<T> {
    return new Promise((resolve, reject) => {
      if (!this.socket?.connected) {
        return reject(new Error('Socket is not connected'));
      }
      this.socket.emit(event, payload, (response: any) => {
        if (response?.error) {
          reject(response.error);
        } else {
          resolve(response as T);
        }
      });
    });
  }
  
  // Hàm này giúp khắc phục lỗi "socket not connected"
  // Sẽ tự động kết nối lại nếu chưa connected, thay vì tự lờ đi
  async safeEmitValidated(event: string, payload: any) {
    if (!this.socket?.connected) {
      console.warn(`[Web Socket] Bị ngắt kết nối. Đang thử kết nối lại để gửi ${event}...`);
      await this.connect();
    }
    
    if (!this.socket?.connected) {
        throw new Error('Mạng bị mất, không thể gửi tín hiệu gọi thoại tại thời điểm này.');
    }
    
    return this.emitWithAck(event, payload);
  }
}

export const socketClient = new SocketClient();

File: apps/mobile-app-fluter/lib/core/socket_client.dart
import 'dart:async';
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'package:flutter/foundation.dart';

// Const endpoint (assuming Web simulator uses 3001)
const String _socketOrigin = 'http://localhost:3001';
const String _chatNamespace = '/chat';

class SocketClient {
  static final SocketClient _instance = SocketClient._internal();
  factory SocketClient() => _instance;
  SocketClient._internal();

  io.Socket? _socket;
  String? _authToken;
  Completer<void>? _connectCompleter;

  Future<void> connect(String token) async {
    if (_connectCompleter != null && _authToken == token) {
      return _connectCompleter!.future;
    }

    _connectCompleter = Completer<void>();

    try {
      if (_socket != null && _authToken != token) {
        _socket!.clearListeners();
        _socket!.disconnect();
        _socket = null;
      }

      if (_socket != null && _authToken == token) {
        if (_socket!.connected) {
          _connectCompleter!.complete();
          return;
        }
        _bindCoreListeners();
        _socket!.connect();
        return;
      }

      _socket = io.io(
        '$_socketOrigin$_chatNamespace',
        io.OptionBuilder()
            .setTransports(['websocket', 'polling'])
            .setAuth({'token': token})
            .enableReconnection()
            .setReconnectionAttempts(20)
            .setReconnectionDelay(800)
            .setTimeout(12000)
            .disableForceNew()
            .disableAutoConnect()
            .build(),
      );
      _authToken = token;

      _bindCoreListeners();
      _socket!.connect();
    } catch (e) {
      if (kDebugMode) {
        print('Failed to init socket: $e');
      }
      _connectCompleter!.completeError(e);
    }

    return _connectCompleter!.future;
  }

  void _bindCoreListeners() {
    if (_socket == null) return;

    _socket!.clearListeners();

    _socket!.onConnect((_) {
      if (kDebugMode) print('Socket connected: ${_socket!.id}');
      if (_connectCompleter != null && !_connectCompleter!.isCompleted) {
        _connectCompleter!.complete();
      }
    });

    _socket!.on('READY', (payload) {
      if (kDebugMode) print('Chat Socket Ready: $payload');
    });

    _socket!.onDisconnect((reason) {
      if (kDebugMode) print('Socket disconnected: $reason');
    });

    _socket!.onConnectError((error) {
      if (kDebugMode) print('Socket connect error: $error');
      if (_connectCompleter != null && !_connectCompleter!.isCompleted) {
        _connectCompleter!.completeError(error);
      }
    });
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.clearListeners();
      _socket!.disconnect();
      _socket = null;
    }
    _authToken = null;
    _connectCompleter = null;
  }

  Future<T?> emitWithAck<T>(String event, dynamic payload) {
    if (_socket == null || !_socket!.connected) {
      return Future.error('Socket is not connected');
    }

    final completer = Completer<T?>();
    _socket!.emitWithAck(event, payload, ack: (dynamic response) {
      if (response != null && response is Map && response.containsKey('error')) {
        completer.completeError(response['error']);
      } else if (response != null && response is Map && response.containsKey('data')) {
        completer.complete(response['data'] as T?);
      } else {
        completer.complete(response as T?);
      }
    });

    return completer.future;
  }

  void on(String event, void Function(dynamic) handler) {
    _socket?.on(event, handler);
  }

  void off(String event, [void Function(dynamic)? handler]) {
    if (handler != null) {
      _socket?.off(event, handler);
    } else {
      _socket?.off(event);
    }
  }
}
